In [85]:
from math import sqrt
from scipy import stats
from pydataset import data
from datetime import datetime


from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
import sklearn.metrics

%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from env import host, user, password
from acquire import get_titanic_data, get_iris_data, get_iris_df, get_titanic_df

#### Model Evaluation Exercises

##### Given the following confusion matrix, evaluate (by hand) the model's performance.


    -|               | pred dog   | pred cat   |
    -|:------------  |-----------:|-----------:|
    -| actual dog    |         46 |         7  |
    -| actual cat    |         13 |         34 |

 In the context of this problem, what is a false positive?
In the context of this problem, what is a false negative?
How would you describe this model?

A false positive in this problem would be that the model predicted cat and it was a dog.

A false negative would be we predicted dog and it was a cat.

#### You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [3]:
ducky = pd.read_csv('c3.csv')
ducky.head(5)

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [27]:
subset = ducky[ducky.actual == 'Defect']
(subset.actual == subset.model1).mean()

0.5

In [30]:
subset = ducky[ducky.actual == 'Defect']
(subset.actual == subset.model2).mean()

0.5625

In [31]:
subset = ducky[ducky.actual == 'Defect']
(subset.actual == subset.model3).mean()

0.8125

    -Model 3 has the best recall prediction rate at 81.25%

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [41]:
ducky.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [50]:
ducky['baseline'] = 'No Defect'

In [51]:
subset = ducky[ducky.model1 == 'Defect']
subset.head() 

actual  model1     model2     model3   baseline
3   No Defect  Defect     Defect     Defect  No Defect
30     Defect  Defect  No Defect     Defect  No Defect
62  No Defect  Defect  No Defect  No Defect  No Defect
65     Defect  Defect     Defect     Defect  No Defect
70     Defect  Defect     Defect     Defect  No Defect

In [52]:
(subset.model1 == subset.actual).mean()

0.8

In [53]:
subset = ducky[ducky.model2 == 'Defect']
subset.head() 

actual     model1  model2     model3   baseline
0  No Defect  No Defect  Defect  No Defect  No Defect
1  No Defect  No Defect  Defect     Defect  No Defect
2  No Defect  No Defect  Defect  No Defect  No Defect
3  No Defect     Defect  Defect     Defect  No Defect
4  No Defect  No Defect  Defect  No Defect  No Defect

In [54]:
(subset.model2 == subset.actual).mean()

0.1

In [55]:
subset = ducky[ducky.model3 == 'Defect']
subset.head() 


actual     model1     model2  model3   baseline
1   No Defect  No Defect     Defect  Defect  No Defect
3   No Defect     Defect     Defect  Defect  No Defect
5   No Defect  No Defect  No Defect  Defect  No Defect
9   No Defect  No Defect  No Defect  Defect  No Defect
13     Defect  No Defect     Defect  Defect  No Defect

In [56]:
(subset.model3 == subset.actual).mean()

0.13131313131313133

    -As we can see when calculating precision for the three models. Model 1 has the highest precision at 80% which would mean it is the least likely to have that false positive outcome. Where c3 gives a hawaii vacation away to someone who didn't have a defective ducky.


##### You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).



At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

In [59]:
paws = pd.read_csv('gives_you_paws.csv')
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

##### Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [60]:
paws.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [62]:
paws['baseline'] = 'dog'
paws.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

##### In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?


In [63]:
(paws.actual == paws.baseline).mean() ### baseline accuracy 

0.6508

In [64]:
(paws.actual == paws.model1).mean() ### model 1 has a higher accuracy than the baseline.

0.8074

In [66]:
(paws.actual == paws.model2).mean() ### model 2 has a lower accuracy than the baseline

0.6304

In [68]:
(paws.actual == paws.model3).mean() ### model 3 has lower accuracy than the baseline

0.5096

In [70]:
(paws.actual == paws.model4).mean() ### model 3 has a higher accuracy than the baseline

0.7426

##### Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?


In [69]:
paws.head(3)

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog

In [71]:
subset = paws[paws.model1 == 'dog']
(subset.model1 == subset.actual).mean()

0.8900238338440586

In [72]:
subset = paws[paws.model2 == 'dog']
(subset.model2 == subset.actual).mean()

0.8931767337807607

In [73]:
subset = paws[paws.model3 == 'dog']
(subset.model3 == subset.actual).mean()

0.6598883572567783

In [74]:
subset = paws[paws.model4 == 'dog']
(subset.model4 == subset.actual).mean()

0.7312485304490948

    -Model 2 has the highest precision in regards to dogs being predicted as a positive. We looked at precision because the dog team that works with dogs photos wouldn't want any cat photos. So we want to look at how good our positive predictions are (dog predictions) and model 2 is slightly better than model 1 at 89.3% and model 3 and 4 are the worst. 
    -Model 1 should be used for phase I for having the highest accuracy. Than model 2 should be used for phase 2 because it has the highest precision.

##### Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?


In [75]:
dogs = paws[paws.actual == 'dog']
dogs.head()

actual model1 model2 model3 model4 baseline
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
5    dog    dog    dog    dog    dog      dog
8    dog    dog    cat    dog    dog      dog

In [77]:
recall1 = (dogs.model1 == dogs.actual).mean()
recall1

0.803318992009834

In [78]:
recall2 = (dogs.model2 == dogs.actual).mean()
recall2

0.49078057775046097

In [79]:
recall3 = (dogs.model3 == dogs.actual).mean()
recall3

0.5086047940995697

In [80]:
recall4 = (dogs.model4 == dogs.actual).mean()
recall4

0.9557467732022127

    -For phase I for the cat team I would use model 1 because it has highest accuracy. For phase II I would use model 4 because it has the highest recall percentage meaning it will prevent the most false negatives in predicting a cat for the cat team when it is actually a dog.

#### Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.



##### Using sklearn.metrics.accuracy_score to calculate model accuracy

In [87]:
accuracy = sklearn.metrics.accuracy_score(paws.actual, paws.model1)
accuracy

0.8074

In [88]:
accuracy2 = sklearn.metrics.accuracy_score(paws.actual, paws.model2)
accuracy2

0.6304

In [89]:
accuracy3 = sklearn.metrics.accuracy_score(paws.actual, paws.model3)
accuracy3

0.5096

In [90]:
accuracy4 = sklearn.metrics.accuracy_score(paws.actual, paws.model4)
accuracy4

0.7426

##### Using skleanr.metrics.precision_score to calculate model precision

In [92]:
precision = sklearn.metrics.precision_score(paws.actual, paws.model1, pos_label='dog')
precision

0.8900238338440586

In [93]:
precision2 = sklearn.metrics.precision_score(paws.actual, paws.model2, pos_label='dog')
precision2

0.8931767337807607

In [94]:
precision3 = sklearn.metrics.precision_score(paws.actual, paws.model3, pos_label='dog')
precision3

0.6598883572567783

In [95]:
precision4 = sklearn.metrics.precision_score(paws.actual, paws.model4, pos_label='dog')
precision4

0.7312485304490948

##### Using sklearn.metrics.recall_score to calculate recall score of models

In [97]:
recall = sklearn.metrics.recall_score(paws.actual, paws.model1, pos_label = 'dog')
recall

0.803318992009834

In [98]:
recall2 = sklearn.metrics.recall_score(paws.actual, paws.model2, pos_label = 'dog')
recall2

0.49078057775046097

In [99]:
recall3 = sklearn.metrics.recall_score(paws.actual, paws.model3, pos_label = 'dog')
recall3

0.5086047940995697

In [100]:
recall4 = sklearn.metrics.recall_score(paws.actual, paws.model4, pos_label = 'dog')
recall4

0.9557467732022127

##### Using sklearn.metrics.classification report to give a report of all score (precision, accuracy, and recall)

In [103]:
class_report = sklearn.metrics.classification_report(paws.actual, paws.model1)
print(f'model 1 report:\n\n{class_report}\n')

model 1 report:

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000




In [104]:
class_report2 = sklearn.metrics.classification_report(paws.actual, paws.model2)
print(f'model 2 report:\n\n{class_report2}\n')

model 2 report:

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000




In [105]:
class_report3 = sklearn.metrics.classification_report(paws.actual, paws.model3)
print(f'model 3 report:\n\n{class_report3}\n')

model 3 report:

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000




In [106]:
class_report4 = sklearn.metrics.classification_report(paws.actual, paws.model4)
print(f'model 4 report:\n\n{class_report4}\n')

model 4 report:

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000


